Core/Core/Dashboard/K5/ViewModel/ImportantDates/K5ImportantDatesViewModel.swift

```diff
@@ -0,0 +1,103 @@
+//
+// This file is part of Canvas.
+// Copyright (C) 2021-present  Instructure, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+import SwiftUI
+
+public class K5ImportantDatesViewModel: ObservableObject {
+
+    @Published public private(set) var importantDates: [K5ImportantDate] = []
+
+    private let env = AppEnvironment.shared
+    private var contexts: [Context] = []
+
+    private lazy var courses = env.subscribe(GetUserCourses(userID: env.currentSession?.userID ?? "")) { [weak self] in
+        self?.coursesUpdated()
+    }
+
+    private lazy var events = env.subscribe(GetCalendarEvents(contexts: contexts, type: .event, importantDates: true)) { [weak self] in
+        self?.eventsUpdated()
+    }
+
+    private lazy var assignments = env.subscribe(GetCalendarEvents(contexts: contexts, type: .assignment, importantDates: true)) { [weak self] in
+        self?.assignmentsUpdated()
+    }
+
+    // MARK: Refresh
+    private var refreshCompletion: (() -> Void)?
+    private var forceRefresh = false
+
+    init() {
+        forceRefresh = true
+        courses.refresh()
+    }
+
+    private func coursesUpdated() {
+        contexts.removeAll()
```

Comment:
Could we write `contexts = courses.map { ... }` instead of this `removeAll` the the `forEach`?